src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs

```diff
@@ -584,5 +588,96 @@ protected HqlTreeNode VisitNewArrayExpression(NewArrayExpression expression)
 			var expressionSubTree = expression.Expressions.Select(exp => VisitExpression(exp)).ToArray();
 			return _hqlTreeBuilder.ExpressionSubTreeHolder(expressionSubTree);
 		}
+
+		private bool IsCastRequired(Expression expression, System.Type toType)
+		{
+			return toType != typeof(object) && IsCastRequired(GetType(expression), TypeFactory.GetDefaultTypeFor(toType));
+		}
+
+		private bool IsCastRequired(IType type, IType toType)
+		{
+			// A type can be null when casting an entity into a base class, in that case we should not cast
+			if (type == null || toType == null || Equals(type, toType))
+			{
+				return false;
+			}
+
+			var sqlTypes = type.SqlTypes(_parameters.SessionFactory);
+			var toSqlTypes = toType.SqlTypes(_parameters.SessionFactory);
+			if (sqlTypes.Length != 1 || toSqlTypes.Length != 1)
+			{
+				return false; // Casting a multi-column type is not possible
+			}
+
+			if (type.ReturnedClass.IsEnum && sqlTypes[0].DbType == DbType.String)
+			{
+				return false; // Never cast an enum that is mapped as string, the type will provide a string for the parameter value
+			}
+
+			return sqlTypes[0].DbType != toSqlTypes[0].DbType;
+		}
+
+		private bool IsCastRequired(System.Type type, string sqlFunctionName)
+		{
+			if (type == typeof(object))
+			{
+				return false;
+			}
+
+			var toType = TypeFactory.GetDefaultTypeFor(type);
+			if (toType == null)
+			{
+				return true; // Fallback to the old behavior
+			}
+
+			var sqlFunction = _parameters.SessionFactory.SQLFunctionRegistry.FindSQLFunction(sqlFunctionName);
+			if (sqlFunction == null)
+			{
+				return true; // Fallback to the old behavior
+			}
+
+			var fnReturnType = sqlFunction.ReturnType(toType, _parameters.SessionFactory);
+			return fnReturnType == null || IsCastRequired(fnReturnType, toType);
+		}
+
+		private IType GetType(Expression expression)
+		{
+			if (!(expression is MemberExpression memberExpression))
+			{
+				return expression.Type != typeof(object)
+					? TypeFactory.GetDefaultTypeFor(expression.Type)
+					: null;
+			}
+
+			// Try to get the mapped type for the member as it may be a non default one
+			var entityName = TryGetEntityName(memberExpression);
+			if (entityName == null)
+			{
+				return TypeFactory.GetDefaultTypeFor(expression.Type); // Not mapped
+			}
+
+			var persister = _parameters.SessionFactory.GetEntityPersister(entityName);
+			var index = persister.EntityMetamodel.GetPropertyIndexOrNull(memberExpression.Member.Name);
+			return !index.HasValue
+				? TypeFactory.GetDefaultTypeFor(expression.Type) // Not mapped
+				: persister.EntityMetamodel.PropertyTypes[index.Value];
+		}
+
+		private string TryGetEntityName(MemberExpression memberExpression)
```

Comment:
Copied from #1996.

Reply:
After doing #2079 I found out that this method does not cover all cases. I will update this method and try to add additional tests that cover the edge cases.

Reply:
This method was improved and exposed for internal usage by #1996. I am thinking to leave this PR as WIP until #1996 get merged in order to replace this method with the new one.